x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/cache/TestUtils.java

```diff
@@ -245,4 +292,62 @@ public void putAsync(
             listener.onResponse(null);
         }
     }
+
+    /**
+     * A {@link FileSystemProvider} that counts the number of times the method {@link FileChannel#force(boolean)} is executed on every
+     * files.
+     */
+    public static class FSyncTrackingFileSystemProvider extends FilterFileSystemProvider {
+
+        private final Map<Path, AtomicInteger> files = new ConcurrentHashMap<>();
+        private final AtomicBoolean failFSyncs = new AtomicBoolean();
+        private final FileSystem delegateInstance;
+        private final Path rootDir;
+
+        public FSyncTrackingFileSystemProvider(FileSystem delegate, Path rootDir) {
+            super("fsynccounting://", delegate);
+            this.rootDir = new FilterPath(rootDir, this.fileSystem);
+            this.delegateInstance = delegate;
+        }
+
+        public FileSystem getDelegateInstance() {
+            return delegateInstance;
+        }
+
+        public void failFSyncs(boolean shouldFail) {
+            failFSyncs.set(shouldFail);
+        }
+
+        public Path resolve(String other) {
+            return rootDir.resolve(other);
+        }
+
+        @Nullable
+        public Integer getNumberOfFSyncs(Path path) {
+            final AtomicInteger counter = files.get(path);
+            return counter != null ? counter.get() : null;
+        }
+
+        @Override
+        public FileChannel newFileChannel(Path path, Set<? extends OpenOption> options, FileAttribute<?>... attrs) throws IOException {
+            final AtomicInteger counter = files.computeIfAbsent(path, p -> new AtomicInteger(0));
+            return new FilterFileChannel(delegate.newFileChannel(toDelegate(path), options, attrs)) {
+
+                @Override
+                public void force(boolean metaData) throws IOException {
+                    if (failFSyncs.get()) {
+                        throw new IOException("simulated");
+                    }
+                    super.force(metaData);
+                    counter.incrementAndGet();
+                }
+            };
+        }
+
+        @Override
+        public void delete(Path path) throws IOException {
```

Comment:
That the tests work with this confused me for a while - but `CacheFile` uses `deleteIfExists`, which delegates to the original file system rather than going here. We need to keep the fsync counters for the check that we no longer fsync after evict to work.

I wonder if this override of `delete` is at all necessary?

Reply:
> That the tests work with this confused me for a while - but `CacheFile` uses `deleteIfExists`, which delegates to the original file system rather than going here.

I agree this is confusing. This delete method is a left over from a previous version of a test that worked differently. I removed the method in e9721c43b8c24cb74475e7ecbd1fb2e4f5feda38. I also noticed a subtility in the `testFSyncOnEvictedFile()` test where the file is not released until the test ends, preventing the file to be deleted from disk. I added some randomization there.